provider/datagen/src/lib.rs

```
@@ -309,36 +310,59 @@ impl DatagenProvider {
+                    Some(iter) => {
+                        let iter = iter.as_ref().map_err(|x| *x)?;
+                        iter.get()
+                    }
+                    None => locale,
+                },
                 metadata,
             };
             let result = self.load_data(key, req);
             match result {
                 Ok(data_response) => {
                     #[allow(clippy::unwrap_used)] // LocaleFallbackProvider populates it
-                    if iter.get().is_empty() && !locale.is_empty() {
-                        log::debug!("Falling back to und: {key}/{locale}");
+                    if let Some(iter) = option_iter.as_ref() {
+                        let iter = iter.as_ref().map_err(|x| *x)?;
+                        if iter.get().is_empty() && !locale.is_empty() {
+                            log::debug!("Falling back to und: {key}/{locale}");
+                        }
                     }
                     return Ok(Some(data_response.take_payload()?));
                 }
                 Err(DataError {
                     kind: DataErrorKind::MissingLocale,
                     ..
                 }) => {
-                    if iter.get().is_empty() {
-                        log::debug!("Could not find data for: {key}/{locale}");
-                        return Ok(None);
-                    }
-                    iter.step();
+                    match option_iter.as_mut() {
+                        Some(iter) => {
+                            let iter = iter.as_mut().map_err(|x| *x)?;
+                            if iter.get().is_empty() {
+                                log::debug!("Could not find data for: {key}/{locale}");
+                                return Ok(None);
+                            }
+                            iter.step();
+                        }
+                        None => (),
+                    };
+                    option_iter.get_or_insert_with(|| {
```

You can return the error here if you don't use `get_or_insert_with`, then you don't have to deal with it in three other locations.